text/0580-destroyables.md

```diff
@@ -0,0 +1,247 @@
+- Calling `unregisterDestructor` with a destructor that is not associated with
+  the object should throw an error.
+
+#### `destroy`
+
+`destroy` initiates the destruction of a destroyable object. It runs all
+associated destructors, and then destroys all children recursively.
+
+Destruction via `destroy()` follows these steps:
+
+1. Mark the destroyable such that `isDestroying(destroyable)` returns `true`
+2. Schedule calling the destroyable's destructors
+3. Call `destroy()` on each of the destroyable's associated children
+4. Schedule setting destroyable such that `isDestroyed(destroyable)` returns `true`
+
+This algorithm results in the entire tree of destroyables being first marked as
+destroying, then having all of their destructors called, and finally all being
+marked as `isDestroyed`. There won't be any in between states where some items
+are marked as `isDestroying` while destroying, while others are not.
+
+Calling `destroy` multiple times on the same destroyable is safe. It will not
+throw an error, and will not take any further action.
+
+Calling `destroy` with a destroyable that has no destructors or associated children
+will not throw an error, and will do nothing.
+
+#### `isDestroying`
+
+Receives a destroyable, and returns `true` if the destroyable has begun
+destroying. Otherwise returns false.
+
+#### `isDestroyed`
+
+Receives a destroyable, and returns `true` if the destroyable has finished
+destroying. Otherwise returns false.
+
+#### `assertDestroyablesDestroyed`
+
+This function asserts that all objects which have associated destructors or
+associated children have been destroyed at the time it is called. It is meant to
+be a low level hook that testing frameworks like `ember-qunit` and `ember-mocha`
+can use to hook into and validate that all destroyables have in fact been
+destroyed.
+
+### Built In Destroyables
+
+The root destroyable of an Ember application will be the instance of the owner.
+All framework managed classes are destroyables, including:
+
+* Components
+* Services
+* Routes
+* Controllers
+* Helpers
+* Modifiers
+
+Any future classes that are added and have a container managed lifecycle should
+also be marked as destroyables.
+
+## How we teach this
+
+This is a generally low level API, meant for usage in the framework and in
+addons. It should be taught primarily through API documentation, and potentially
+through an in-depth guide.
+
+## Drawbacks
+
+- Adds another destruction API which may conflict with the existing destruction
+  hooks. Since this is a low-level API, it shouldn't be too problematic - most
+  users will be guided toward using the standard lifecycle hooks, and this API
+  will exist for libraries like `ember-concurrency` and `ember-lifeline`.
+
+## Alternatives
+
+- Continue using existing lifecycle hooks for public API, and don't provide an
+  independent API.
+
+## Unresolved questions
+
+- Should `willDestroy` and `didDestroy` (e.g. mark-and-sweep style semantics) be
```

Comment:
Just throwing a thought: Will this be a footgun? Seems like first thing I would try if `willDestroy` get deprecated and codemods are not working for me (for whatever reason):
```js
registerDestructor(this, this.willDestroy);
```

Reply:
This specifically was referencing _synchronous_ `willDestroy` semantics, which are a footgun for performance, but are necessary for things like running code before a component is removed from the DOM (e.g. `willDestroyElement`). Generally, you should do the minimal amount of work synchronously, and schedule the rest, but often this doesn't happen. I'll reword this to make that more clear.

Reply:
FWIW, the current semantics _would_ be exactly the same as `willDestroy` is currently.